src/compiler/checker.ts

```diff
@@ -25198,7 +25202,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const constraint = getConstraintOfTypeParameter(inference.typeParameter);
             if (constraint) {
                 const instantiatedConstraint = instantiateType(constraint, context.nonFixingMapper);
-                if (!inferredType || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType))) {
+                if (!inferredType || inferredType === wildcardType || !context.compareTypes(inferredType, getTypeWithThisArgument(instantiatedConstraint, inferredType))) {

```

Comment:
This change might be best reviewed in isolated by reviewing the first commit: https://github.com/microsoft/TypeScript/pull/53996/commits/7ab8679aab721587d5207456cfc11ae7230912ac